ocaml/testsuite/tests/typing-layouts-err-msg/test.ml

```diff
@@ -178,3 +178,80 @@ Error: This expression has type t_void but an expression was expected of type
        But the layout of t_void must be a sublayout of value, because
          of layout requirements from an imported definition.
 |}]
+
+type ('a : value) t_v = 'a
+type ('a : void) t_vv = 'a
+let f x: 'a t_v = x
+let f2 = f
+let () = ignore (f2 (assert false : 'a t_vv))
+
+[%%expect{|
+type 'a t_v = 'a
+type ('a : void) t_vv = 'a
+val f : 'a t_v -> 'a t_v = <fun>
+val f2 : 'a t_v -> 'a t_v = <fun>
+Line 5, characters 20-44:
+5 | let () = ignore (f2 (assert false : 'a t_vv))
+                        ^^^^^^^^^^^^^^^^^^^^^^^^
+Error: This expression has type 'a t_vv = ('a : void)
+       but an expression was expected of type 'b t_v = ('b : value)
+       The layout of 'a is value, because
+         of definition of f2 at Line 4, characters 9-10.
+       But the layout of 'a must overlap with void, because
+         of definition of t_vv at Line 2, characters 0-26.
+|}]
+
+type ('a : value) t_v = 'a
+type ('a : void) t_v1 = 'a
+type 'a t_v2 = 'a t_v1
+let f x: 'a t_v = x
+
+let () = ignore (f (assert false : 'a t_v2))
+
+[%%expect{|
+type 'a t_v = 'a
+type ('a : void) t_v1 = 'a
+type ('a : void) t_v2 = 'a t_v1
+val f : 'a t_v -> 'a t_v = <fun>
+Line 6, characters 19-43:
+6 | let () = ignore (f (assert false : 'a t_v2))
+                       ^^^^^^^^^^^^^^^^^^^^^^^^
+Error: This expression has type 'a t_v2 = ('a : void)
+       but an expression was expected of type 'b t_v = ('b : value)
+       The layout of 'a is value, because
+         of definition of f at Line 4, characters 6-19.
+       But the layout of 'a must overlap with void, because
+         of definition of t_v2 at Line 3, characters 0-22.
+|}]
+
+
+module type S1 = sig
+  type ('a : void) t = 'a
+end
+
+module type S2 = S1
+
+module M : S2 = struct
+  type 'a t = 'a
+end
+
+[%%expect{|
+module type S1 = sig type ('a : void) t = 'a end
+module type S2 = S1
+Lines 7-9, characters 16-3:
+7 | ................struct
+8 |   type 'a t = 'a
+9 | end
+Error: Signature mismatch:
+       Modules do not match: sig type 'a t = 'a end is not included in S2
+       Type declarations do not match:
+         type 'a t = 'a
+       is not included in
+         type ('a : void) t = 'a
+       The type ('a : value) is not equal to the type ('a0 : void)
+       because their layouts are different.
+       The layout of 'a is value, because
```

Comment:
I think in a module mismatch error, the histories are always going to be boring. Can we leave them off here?

Reply:
That's a good idea, but don't see a straightforward way of implementing it. There has to be some kind of global state to toggle history on/off or it has to be threaded through all the recursive functions in `includemod_errorprint.ml`.

Reply:
OK.